File: packages/schema/generate.mjs
/**
 * Merge OpenAPI fragments by key scope. Twin of generate.py so the dashboard
 * and the worker emit the same contract (PLAN.md §10). Do not fork the merge
 * rules here without also changing generate.py.
 *
 * Fragments are imported (not read at runtime) so Netlify's esbuild bundle
 * includes them. Sort order matches generate.py's glob of filenames.
 */
import chat from "./fragments/chat.json" with { type: "json" };
import common from "./fragments/common.json" with { type: "json" };
import embeddings from "./fragments/embeddings.json" with { type: "json" };
import health from "./fragments/health.json" with { type: "json" };
import jobs from "./fragments/jobs.json" with { type: "json" };
import models from "./fragments/models.json" with { type: "json" };
import openapi from "./fragments/openapi.json" with { type: "json" };
import usage from "./fragments/usage.json" with { type: "json" };
import scopesJson from "./scopes.json" with { type: "json" };

const FRAGMENTS = [chat, common, embeddings, health, jobs, models, openapi, usage];
const FORBIDDEN_COMPOSITION = ["oneOf", "anyOf", "allOf", "not"];
const APEX_OPERATION_ID = /^[A-Za-z][A-Za-z0-9_]*$/;
const SALESFORCE_MEDIA_TYPE = "application/json";

export function loadScopes() {
  return scopesJson;
}

function loadFragments() {
  return FRAGMENTS;
}

function scopeAllows(fragment, scopes, salesforce) {
  if (salesforce && fragment["x-cloudiator-salesforce"] === false) return false;
  if (fragment["x-cloudiator-always"]) return true;
  const needed = fragment["x-cloudiator-scope"] || [];
  return needed.length > 0 && needed.every((item) => scopes.has(item));
}

function mergeDict(dst, src) {
  for (const [key, value] of Object.entries(src)) {
    if (key.startsWith("x-cloudiator-")) continue;
    if (
      value &&
      typeof value === "object" &&
      !Array.isArray(value) &&
      dst[key] &&
      typeof dst[key] === "object" &&
      !Array.isArray(dst[key])
    ) {
      mergeDict(dst[key], value);
    } else {
      dst[key] = structuredClone(value);
    }
  }
}

function collectComposition(node, found) {
  if (node && typeof node === "object" && !Array.isArray(node)) {
    for (const [key, value] of Object.entries(node)) {
      if (FORBIDDEN_COMPOSITION.includes(key)) found.push(key);
      collectComposition(value, found);
    }
  } else if (Array.isArray(node)) {
    for (const item of node) collectComposition(item, found);
  }
}

export function compositionKeys(doc) {
  const found = [];
  collectComposition(doc, found);
  return found;
}

function operationIds(doc) {
  const ids = [];
  for (const pathItem of Object.values(doc.paths || {})) {
    if (!pathItem || typeof pathItem !== "object") continue;
    for (const [method, op] of Object.entries(pathItem)) {
      if (method.startsWith("x-") || !op || typeof op !== "object") continue;
      if (op.operationId) ids.push(op.operationId);
    }
  }
  return ids;
}

function walkSchemas(node, path = "") {
  const out = [];
  if (node && typeof node === "object" && !Array.isArray(node)) {
    if ("type" in node || "$ref" in node || "properties" in node) {
      out.push([path, node]);
    }
    for (const [key, value] of Object.entries(node)) {
      out.push(...walkSchemas(value, path ? `${path}.${key}` : key));
    }
  } else if (Array.isArray(node)) {
    node.forEach((item, index) => {
      out.push(...walkSchemas(item, `${path}[${index}]`));
    });
  }
  return out;
}

function mediaTypes(doc) {
  const found = [];
  for (const pathItem of Object.values(doc.paths || {})) {
    if (!pathItem || typeof pathItem !== "object") continue;
    for (const operation of Object.values(pathItem)) {
      if (!operation || typeof operation !== "object") continue;
      const bodies = [operation.requestBody || {}];
      bodies.push(...Object.values(operation.responses || {}));
      for (const body of bodies) {
        if (body && typeof body === "object") {
          found.push(...Object.keys(body.content || {}));
        }
      }
    }
  }
  return found;
}

function assertSalesforceSafe(doc) {
  const bad = compositionKeys(doc);
  if (bad.length) {
    throw new Error(`Salesforce OAS forbids composition keywords: ${[...new Set(bad)].sort()}`);
  }
  const ids = operationIds(doc);
  if (ids.length !== new Set(ids).size) {
    throw new Error("operationId values must be unique");
  }
  for (const oid of ids) {
    if (!APEX_OPERATION_ID.test(oid)) {
      throw new Error(`operationId is not Apex-safe: ${JSON.stringify(oid)}`);
    }
  }
  for (const mediaType of mediaTypes(doc)) {
    if (mediaType !== SALESFORCE_MEDIA_TYPE) {
      throw new Error(`Salesforce OAS is ${SALESFORCE_MEDIA_TYPE} only, found ${JSON.stringify(mediaType)}`);
    }
  }
  for (const [where, schema] of walkSchemas((doc.components || {}).schemas || {})) {
    if ("additionalProperties" in schema) {
      throw new Error(`free-form additionalProperties at ${where}`);
    }
    if (schema.format === "binary") {
      throw new Error(`format: binary is not importable, at ${where}`);
    }
    if ("$ref" in schema) {
      const ref = schema.$ref;
      if (!String(ref).startsWith("#/")) {
        throw new Error(`external $ref at ${where}: ${ref}`);
      }
      continue;
    }
    if (schema.type === "object" && !schema.properties) {
      throw new Error(`object schema without properties at ${where}`);
    }
    if (schema.type === "array" && !schema.items) {
      throw new Error(`array schema without items at ${where}`);
    }
  }
}

export function buildOpenApi(
  scopes,
  { target = null, publicBaseUrl = "https://api.example.com", version = "0.1.0" } = {},
) {
  const scopeSet = new Set(scopes);
  const isSalesforce = String(target || "").toLowerCase() === "salesforce";
  const merged = { paths: {}, components: { schemas: {} } };
  for (const fragment of loadFragments()) {
    if (!scopeAllows(fragment, scopeSet, isSalesforce)) continue;
    mergeDict(merged, fragment);
  }

  const doc = {
    openapi: isSalesforce ? "3.0.3" : "3.1.0",
    info: {
      title: "Cloudiator",
      version,
      description: "Scoped OpenAI-compatible appliance API",
    },
    servers: [{ url: publicBaseUrl.replace(/\/$/, "") }],
    paths: merged.paths || {},
    components: merged.components || {},
    security: [{ bearerAuth: [] }],
  };
  const health = doc.paths["/v1/health"]?.get;
  if (health && typeof health === "object") {
    health.security = [];
  }
  if (isSalesforce) assertSalesforceSafe(doc);
  return doc;
}

export function dumpOpenApi(scopes, options) {
  return `${JSON.stringify(buildOpenApi(scopes, options), null, 2)}\n`;
}

File: packages/schema/generate.py
"""Merge OpenAPI fragments by key scope. No Ollama, no Neon, no FastAPI."""

from __future__ import annotations

import copy
import json
from pathlib import Path
from typing import Any, Iterable, Mapping

FRAGMENTS_DIR = Path(__file__).resolve().parent / "fragments"
SCOPES_PATH = Path(__file__).resolve().parent / "scopes.json"

FORBIDDEN_COMPOSITION = ("oneOf", "anyOf", "allOf", "not")
APEX_OPERATION_ID = __import__("re").compile(r"^[A-Za-z][A-Za-z0-9_]*$")


def load_scopes() -> dict[str, Any]:
    return json.loads(SCOPES_PATH.read_text())


def _load_fragments() -> list[dict[str, Any]]:
    files = sorted(FRAGMENTS_DIR.glob("*.json"))
    if not files:
        raise FileNotFoundError(f"no OpenAPI fragments in {FRAGMENTS_DIR}")
    return [json.loads(path.read_text()) for path in files]


def _scope_allows(fragment: Mapping[str, Any], scopes: set[str]) -> bool:
    if fragment.get("x-cloudiator-always"):
        return True
    needed = fragment.get("x-cloudiator-scope") or []
    return bool(needed) and set(needed).issubset(scopes)


def _merge_dict(dst: dict[str, Any], src: Mapping[str, Any]) -> None:
    for key, value in src.items():
        if key.startswith("x-cloudiator-"):
            continue
        if isinstance(value, Mapping) and isinstance(dst.get(key), dict):
            _merge_dict(dst[key], value)
        else:
            dst[key] = copy.deepcopy(value)


def _collect_composition(node: Any, found: list[str]) -> None:
    if isinstance(node, dict):
        for key, value in node.items():
            if key in FORBIDDEN_COMPOSITION:
                found.append(key)
            _collect_composition(value, found)
    elif isinstance(node, list):
        for item in node:
            _collect_composition(item, found)


def composition_keys(doc: Mapping[str, Any]) -> list[str]:
    found: list[str] = []
    _collect_composition(doc, found)
    return found


def _operation_ids(doc: Mapping[str, Any]) -> list[str]:
    ids: list[str] = []
    for path_item in (doc.get("paths") or {}).values():
        if not isinstance(path_item, dict):
            continue
        for method, op in path_item.items():
            if method.startswith("x-") or not isinstance(op, dict):
                continue
            oid = op.get("operationId")
            if oid:
                ids.append(oid)
    return ids


def _assert_salesforce_safe(doc: Mapping[str, Any]) -> None:
    bad = composition_keys(doc)
    if bad:
        raise ValueError(f"Salesforce OAS forbids composition keywords: {sorted(set(bad))}")
    ids = _operation_ids(doc)
    if len(ids) != len(set(ids)):
        raise ValueError("operationId values must be unique")
    for oid in ids:
        if not APEX_OPERATION_ID.match(oid):
            raise ValueError(f"operationId is not Apex-safe: {oid!r}")


def build_openapi(
    scopes: Iterable[str],
    *,
    target: str | None = None,
    public_base_url: str = "https://api.example.com",
    version: str = "0.1.0",
) -> dict[str, Any]:
    """Return an OpenAPI document for the given key scopes.

    target='salesforce' emits OpenAPI 3.0.3 with the restricted subset
    External Services will import (PLAN.md §10).
    """
    scope_set = set(scopes)
    merged: dict[str, Any] = {"paths": {}, "components": {"schemas": {}}}
    for fragment in _load_fragments():
        if not _scope_allows(fragment, scope_set):
            continue
        _merge_dict(merged, fragment)

    is_salesforce = (target or "").lower() == "salesforce"
    doc: dict[str, Any] = {
        "openapi": "3.0.3" if is_salesforce else "3.1.0",
        "info": {
            "title": "Cloudiator",
            "version": version,
            "description": "Scoped OpenAI-compatible appliance API",
        },
        "servers": [{"url": public_base_url.rstrip("/")}],
        "paths": merged.get("paths") or {},
        "components": merged.get("components") or {},
        "security": [{"bearerAuth": []}],
    }
    if "/v1/health" in doc["paths"]:
        health = doc["paths"]["/v1/health"].get("get")
        if isinstance(health, dict):
            health["security"] = []

    if is_salesforce:
        _assert_salesforce_safe(doc)
    return doc


def dump_openapi(scopes: Iterable[str], **kwargs: Any) -> str:
    return json.dumps(build_openapi(scopes, **kwargs), indent=2) + "\n"


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Build a filtered OpenAPI document")
    parser.add_argument(
        "--scopes",
        default="chat",
        help="comma-separated capabilities (default: chat)",
    )
    parser.add_argument("--target", default="", help="salesforce for 3.0.3 subset")
    parser.add_argument("--base-url", default="https://api.example.com")
    args = parser.parse_args()
    scopes = [s.strip() for s in args.scopes.split(",") if s.strip()]
    print(
        dump_openapi(
            scopes,
            target=args.target or None,
            public_base_url=args.base_url,
        ),
        end="",
    )
